### PR TITLE
feat(flow): add `eval_batch` for ScalarExpr

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

add `eval_batch` for flow's `ScalarExpr`, which can be use later to optimize performance for flow and reduce small object alloc/dealloc(which is found to be a very large cost on cpu time when profiling flow tasks).

The major change of logic happens in the following files:
- `src/flow/src/expr/func.rs` include three `eval_batch` function
- `src/flow/src/expr/scalar.rs` 's `ScalarExpr::eval_batch` function
- `src/flow/src/expr/df_func.rs`'s `eval_batch` function

and other changes are just moving file around and remove unused imports

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

**Comment by fengjiachun** [comment]
@discord9 Good job, could you show us the performance analysis comparison before and after this PR to clearly see the optimization effect?

**Comment by discord9** [comment]
> @discord9 Good job, could you show us the performance analysis comparison before and after this PR to clearly see the optimization effect?

It would still require some more change of code to make MapFilterProject Operator to use batching process, which I'm working on adding it,  will include performance analysis in later PR

File: src/flow/src/expr/scalar.rs [hunk 1]
```diff
@@ -428,6 +260,92 @@ impl ScalarExpr {
         }
     }
 
+    pub fn eval_batch(&self, batch: &Batch) -> Result<VectorRef, EvalError> {
+        match self {
+            ScalarExpr::Column(i) => Ok(batch.batch()[*i].clone()),
+            ScalarExpr::Literal(val, dt) => Ok(Helper::try_from_scalar_value(
+                val.try_to_scalar_value(dt).context(DataTypeSnafu {
+                    msg: "Fail to convert literal to scalar value",
```
Comment (fengjiachun) [inline src/flow/src/expr/scalar.rs:268]:
```suggestion
                    msg: "Failed to convert literal to scalar value",
```

File: src/flow/src/expr/scalar.rs [hunk 2]
```diff
@@ -428,6 +260,92 @@ impl ScalarExpr {
...
+                })?,
+                batch.row_count(),
+            )
+            .context(DataTypeSnafu {
+                msg: "Fail to convert scalar value to vector ref when parsing literal",
+            })?),
+            ScalarExpr::CallUnmaterializable(_) => OptimizeSnafu {
+                reason: "Can't eval unmaterializable function".to_string(),
```
Comment (fengjiachun) [inline src/flow/src/expr/scalar.rs:276]:
```suggestion
                reason: "Can't eval unmaterializable function",
```

File: src/flow/src/expr/func.rs [hunk 3]
```diff
@@ -221,6 +225,129 @@ impl UnaryFunc {
         }
     }
 
+    pub fn eval_batch(&self, batch: &Batch, expr: &ScalarExpr) -> Result<VectorRef, EvalError> {
+        let arg_col = expr.eval_batch(batch)?;
+        match self {
+            &Self::Not => {
```
Comment (fengjiachun) [inline src/flow/src/expr/func.rs:231]:
```suggestion
            Self::Not => {
```

File: src/flow/src/expr/func.rs [hunk 4]
```diff
@@ -826,6 +1062,54 @@ impl VariadicFunc {
...
+                    expected: ConcreteDataType::boolean_datatype(),
+                    actual: ConcreteDataType::from_arrow_type(first.data_type()),
+                }
+            })?
+            .clone();
+
+        for right in iter {
+            let right = right
+                .as_any()
+                .downcast_ref::<arrow::array::BooleanArray>()
```
Comment (fengjiachun) [inline src/flow/src/expr/func.rs:1093]:
```suggestion
                .downcast_ref::<BooleanArray>()
```

**Comment by fengjiachun** [comment]
```
error: unused import: `common_telemetry::debug`
  --> src/flow/src/expr/func.rs:22:5
   |
22 | use common_telemetry::debug;
   |     ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D unused-imports` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(unused_imports)]`
```

CI failed, cc @discord9

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/flow/src/expr/scalar.rs [hunk 5]
```diff
@@ -428,6 +261,89 @@ impl ScalarExpr {
...
+                        .batch()
+                        .iter()
+                        .map(|v| v.clone().slice(idx, 1))
+                        .collect::<Vec<_>>();
+                    let input_batch = Batch::new(input_vectors, 1);
+                    let res: VectorRef = match cond {
+                        Some(true) => then.eval_batch(&input_batch)?,
+                        Some(false) => els.eval_batch(&input_batch)?,
+                        None => Arc::new(NullVector::new(1)),
+                    };
```
Comment (zhongzc) [inline src/flow/src/expr/scalar.rs:309]: Either call `eval` for one value directly, or separate the values ​​whose cond is true/false/none and call `eval_batch` and then assemble them back?

**Review by zhongzc — APPROVED** [review]

**Pull request merged**